packages/contracts/contracts/maci/Tally.sol

```diff
@@ -0,0 +1,305 @@
+    uint256 amountLength = amounts.length;
+    uint256 totalFunds = token.balanceOf(address(this));
+    uint256 sum = 0;
+
+    for (uint256 index = 0; index < amountLength; ) {
+      uint256 amount = amounts[index];
+      address receiver = receivers[index];
+      sum += amount;
+
+      if (sum > totalFunds) {
+        revert InvalidWithdrawal();
+      }
+
+      if (amount > 0) {
+        token.safeTransfer(receiver, amount);
+      }
+
+      unchecked {
+        index++;
+      }
+    }
+  }
+
+  /// @inheritdoc IPayoutStrategy
+  function totalAmount() public view override returns (uint256) {
+    return token.balanceOf(address(this));
+  }
+
+  /// @inheritdoc TallyBase
+  function addTallyResults(TallyBase.AddTallyResultsArgs calldata args) public override isInitialized onlyOwner {
+    if (recipientCount == 0) {
+      recipientCount = registry.recipientCount();
+    }
+
+    if (recipientCount == totalTallyResults) {
+      revert TooManyResults();
+    }
+
+    super.addTallyResults(args);
+
+    if (recipientCount < totalTallyResults) {
+      revert TooManyResults();
+    }
+  }
+
+  /// @inheritdoc TallyBase
+  function addTallyResult(
+    uint256 voteOptionIndex,
+    uint256 tallyResult,
+    uint256[][] calldata tallyResultProof,
+    uint256 tallyResultSalt,
+    uint256 spentVoiceCreditsHash,
+    uint256 perVOSpentVoiceCreditsHash,
+    uint8 voteOptionTreeDepth
+  ) internal override {
+    super.addTallyResult(
+      voteOptionIndex,
+      tallyResult,
+      tallyResultProof,
+      tallyResultSalt,
+      spentVoiceCreditsHash,
+      perVOSpentVoiceCreditsHash,
+      voteOptionTreeDepth
+    );
+
+    totalVotesSquares += tallyResult ** 2;
+  }
+
+  /// @inheritdoc IPayoutStrategy
+  function claim(IPayoutStrategy.Claim calldata params) public override isInitialized whenNotPaused afterTallying {
+    if (alpha == 0) {
+      alpha = calculateAlpha(totalAmount());
+    }
+
+    uint256 tallyResult = tallyResults[params.index].value;
+    uint256 amount = getAllocatedAmount(params.index, params.voiceCreditsPerOption);
+
+    IRecipientRegistry.Recipient memory recipient = registry.getRecipient(params.index);
+
+    if (claimed[params.index]) {
+      revert AlreadyClaimed();
+    }
+
+    claimed[params.index] = true;
+
+    bool isValid = verifyTallyResult(
+      params.index,
+      tallyResult,
+      params.tallyResultProof,
+      params.tallyResultSalt,
+      params.voteOptionTreeDepth,
+      params.spentVoiceCreditsHash,
+      params.perVOSpentVoiceCreditsHash
+    );
+
+    if (!isValid) {
+      revert InvalidTallyVotesProof();
+    }
+
+    token.safeTransfer(recipient.recipient, amount);
+  }
+
+  /// @notice Get allocated token amounts (without verification).
+  /// @param voiceCreditsPerOptions The voice credit options received for recipient
+  function getAllocatedAmounts(
+    uint256[] calldata voiceCreditsPerOptions
+  ) public view afterTallying returns (uint256[] memory amounts) {
```

Comment:
don't think there's need to have afterTallying here, when called by other functions they also have the same modifier.

Reply:
ballots tallied not necessarily means that the results have been added (sure in our version with our scripts it does cause the scripts do both operations in once, but they are not necessarily done together, nor would be true in the case of multiple batches of results and the call made before the second batch is added). Either way not a problem, just a small suggestion

Reply:
So I'd keep this one. There is also a check if we don't have full results uploaded.